copy_shard.c

```diff
@@ -0,0 +1,139 @@
+/*-------------------------------------------------------------------------
+ *
+ * copy_shard.c
+ *
+ * This file contains functions to copy a shard's data from a healthy
+ * placement.
+ *
+ * Copyright (c) 2014, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "c.h"
+#include "fmgr.h"
+#include "miscadmin.h"
+#include "postgres_ext.h"
+
+#include "copy_shard.h"
+#include "distribution_metadata.h"
+#include "pg_shard.h"
+
+#include <string.h>
+
+#include "access/heapam.h"
+#include "access/htup.h"
+#include "access/tupdesc.h"
+#include "executor/tuptable.h"
+#include "lib/stringinfo.h"
+#include "nodes/pg_list.h"
+#include "storage/lock.h"
+#include "utils/builtins.h"
+#include "utils/elog.h"
+#include "utils/palloc.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
+#include "utils/tuplestore.h"
+
+
+/* local function forward declarations */
+static bool CopyDataFromTupleStoreToRelation(Tuplestorestate *tupleStore,
+											 Relation relation);
+
+
+/* declarations for dynamic loading */
+PG_FUNCTION_INFO_V1(worker_copy_shard_placement);
```

Comment:
What if we kept this function as part of repair_shards.h/.c for now?

I realize we could conceptually use this UDF in places other than shard repairs. For now, our implementation in this file is 100-200 lines. Therefore, I have a slight preference for keeping this in repair_shards, and consider splitting this once we have 3 separate callers to this function. Any strong objections?